src/helper/runner.cpp

```diff
@@ -36,21 +37,50 @@ network::base_acceptor::ptr acceptor_from_config(const config::config &cfg)
 } // namespace
 
 runner::runner(const config::config &cfg)
-    : cfg_(cfg), engine_pool_{std::make_shared<engine_pool>()},
-      acceptor(acceptor_from_config(cfg))
-{
-}
+    : runner(cfg, acceptor_from_config(cfg)) {}
 
 runner::runner(const config::config &cfg, network::base_acceptor::ptr &&acceptor)
     : cfg_(cfg), engine_pool_{std::make_shared<engine_pool>()},
-      acceptor(std::move(acceptor))
+      acceptor_(std::move(acceptor)),
+      idle_timeout_(cfg.get<unsigned>("runner_idle_timeout"))
 {
+    try {
+        acceptor_->set_accept_timeout(1min);
+    } catch (const std::exception &e) {
+        // Not a critical error, we should continue
+        SPDLOG_WARN("Failed to set runner timeout: {}", e.what());
+    }
 }
 
 void runner::run() {
     try {
+        auto last_idle = std::chrono::steady_clock::now();
+        SPDLOG_INFO("Running");
         while (running_) {
-            auto socket = acceptor->accept();
+            network::base_socket::ptr socket;
+            try {
+                socket = acceptor_->accept();
+            } catch (const timeout_error &e) {
+                // Prevent a missed EINTR
```

Comment:
how can an eintr cause this exception to be raised?

Reply:
This exception can't be raised from `EINTR`. However if there is a signal while the timeout exception is being thrown, the signal handler will call `exit` on the runner which will set `running_` to false. The comment arises from the fact that if we had been in `accept`, there would've been an `EINTR` which would've caused this function to finalise.